doc/man/man4/cd_daed.4.man
'\" rtp
.\" vim: ft=nroff sw=4 noet nocin nosi com=b\:.\\\" fo+=tcqlorn tw=77
.\" =========================================================================
.\"
.\" @(#) doc/man/man4/cd_daed.4.man
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2008-2011  Monavacon Limited <http://www.monavacon.com/>
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\" Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to copy, distribute and/or modify this manual under
.\" the terms of the GNU Free Documentation License, Version 1.3 or any later
.\" version published by the Free Software Foundation; with no Invariant
.\" Sections, no Front-Cover Texts, and no Back-Cover Texts.  A copy of the
.\" license is included in the section entitled "GNU Free Documentation
.\" License".
.\"
.\" Permission to use, copy and distribute this manual without modification,
.\" for any purpose and without fee or royalty is hereby granted, provided
.\" that both the above copyright notice and this permission notice appears
.\" in all copies and that the name of OpenSS7 Corporation not be used in
.\" advertising or publicity pertaining to distribution of this documentation
.\" or its contents without specific, written prior permission.  OpenSS7
.\" Corporation makes no representation about the suitability of this manual
.\" for any purpose.  It is provided "as is" without express or implied
.\" warranty.
.\"
.\" Permission is granted to process this file through groff and print the
.\" results, provided the printed document carries a copying permission
.\" notice identical to this one except for the removal of this paragraph
.\" (this paragraph not being relevant to the printed manual).
.\"
.\" OPENSS7 CORPORATION DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS MANUAL
.\" INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
.\" PARTICULAR PURPOSE, NON-INFRINGEMENT, OR TITLE; THAT THE CONTENTS OF THE
.\" DOCUMENT ARE SUITABLE FOR ANY PURPOSE, OR THAT THE IMPLEMENTATION OF SUCH
.\" CONTENTS WILL NOT INFRINGE ON ANY THIRD PARTY PATENTS, COPYRIGHTS,
.\" TRADEMARKS OR OTHER RIGHTS.  IN NO EVENT SHALL OPENSS7 CORPORATION BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, SPECIAL OR CONSEQUENTIAL DAMAGES OR ANY
.\" DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER
.\" IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
.\" OUT OF OR IN CONNECTION WITH ANY USE OF THIS DOCUMENT OR THE PERFORMANCE
.\" OR IMPLEMENTATION OF THE CONTENTS THEREOF.
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this manual
.\" page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from the
.\" use of the information contained herein.  The author(s) may not have
.\" taken the same level of care in the production of this manual, which is
.\" licensed free of charge, as they might when working professionally.  The
.\" author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by the
.\" source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
.\" behalf of the U.S. Government ("Government"), the following provisions
.\" apply to you.  If the Software is supplied by the Department of Defense
.\" ("DoD"), it is classified as "Commercial Computer Software" under
.\" paragraph 252.227-7014 of the DoD Supplement to the Federal Acquisition
.\" Regulations ("DFARS") (or any successor regulations) and the Government
.\" is acquiring only the license rights granted herein (the license rights
.\" customarily provided to non-Government users).  If the Software is
.\" supplied to any unit or agency of the Government other than DoD, it is
.\" classified as "Restricted Computer Software" and the Government's rights
.\" in the Software are defined in paragraph 52.227-19 of the Federal
.\" Acquisition Regulations ("FAR") (or any successor regulations) or, in the
.\" cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the
.\" FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database openss7.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
search-truncate 40
search-ignore CGIQOSTU
.R2
.so openss7.macros
.\"
.\"
.TH CD-DAED 4 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Devices"
.\"
.\"
.SH NAME
.B cd_daed
\- Delimination Alignment and Error Detection
.\"
.\"
.SH SYNOPSIS
.PP
.B #include <sys/stropts.h>
.br
.B #include <sys/cdi.h>
.br
.B #include <sys/cdi_daed.h>
.br
.B #include <sys/daed_ioctl.h>
.HP
.BR "int " ch_stream " = open(" ch_device ", " flags );
.HP
.BI "int " retval " = ioctl(" ch_device ", " \fRI_PUSH\fI ", " \fR\*(lqcd-daed\*(rq\fI );
.\"
.\"
.SH DESCRIPTION
.PP
.B CD-DAED
is an implementation of the Delimination Aligment and Error Detection (DAED) Procedures as specified
in
.B ITU-T Q.703
.[
q703
.]
and
.BR "ANSI T1.111.3" .
.[
t1.111
.]
.B CD-DAED
is intended to be used with the
.BR sl-mux (4)
multiplexing driver.
.\"
.\"
.PP
.B CD-DAED
is implementated as a pushable
.IR STREAMS (4)
module.  The
.B CD-DAED
module can be pushed over a stream conforming to the Channel Interface (CHI) as described in
.BR chi (7).
The
.B CD-DAED
module provides DAED acces to the bit stream from the channel provided by the
.BR chi (7)
Stream below it, by performing HDLC and DAED functions on the raw bit stream.  The upper interface
provided by the
.B CD-DAED
module is the Communications Device Interface (CDI) as described in
.BR cdi (7).
.PP
.B CD-DAED
Streams can be linked under
.BR cd (4)
drivers
.BR sl-mux (4),
using the
.IR I_LINK " or " I_PLINK
commands of
.BR streamio (7),
to provide a complete Communications Device that can then be linked under a
.BR dl (4)
driver using the
.IR I_LINK " or " I_PLINK
commands of
.BR streamio (7),
to provide data link services to the Layer 3 protocol.  This is normally performed, as required, by
the SS7 Configuration Daemon,
.BR ss7confd (8).
.\"
.\"
.SH NOTICES
.PP
.B CD-DAED
has been implemented as a pushable module to ease the development of SS7 Communications Device
and Data Link drivers for various hardware cards.  All that is required is for the card to provide a
Channel Interface (CHI),
.BR chi (7),
and push the
.BR cd-daed
.BR hdlc (4)
and
.BR dl (4)
modules to provide a complete and compliant SS7 Signalling Link.
.\"
.\"
.PP
.B CD-DAED
is implemented internally as a soft-HDLC using host-based table lookups.  While this is fairly
efficient, devices that are capable of performing this function in hardware should provide the
Communications Device Interface (CDI),
.BR cdi (7),
directly.
An example of a device that does not include HDLC is the
.BR x400p (4)
driver.  An examle of one that does, is the
.BR acb56 (4)
driver.
.\"
.\"
.PP
.B CD-DAED
modules are not normally pushed or accessed directly by user-level programs.
The
.B CD-DAED
module is used directly by some test and monitoring programs.
.\"
.\"
.SH IOCTLS
.PP
The
.B CD-DAED
module responsds to Local Management Interface (LMI),
.BR lmi_ioctl (4),
IO controls as interpreted for
.B CD-DAED
operations as described in
.BR daed_ioctl (4).
.\"
.\"
.SH FILES
.PP
.BR <ss7/cdi.h> ,
.BR <ss7/cdi_daed.h> ,
.BR <ss7/daed_ioctl.h> .
.\"
.\"
.SH MODULES
.PP
.RI streams- cd_daed .o
.\"
.\"
.SH "SEE ALSO"
.PP
.BR cdi (7),
.BR dl (4),
.BR daed_ioctl (4),
.BR ss7confd (8).
.\"
.\"
.SH VERSIONS
.PP
The
.B CD-DAED
module conforms to the CDI Specifications, Revision 1.0.
.[
cdi revision @PACKAGE_VERSION@ draft @PACKAGE_RELEASE@
.]
.PP
The
.B CD-DAED
module does not appear until
.B strchan
release 0.9.2.
.\"
.\"
.\"
.[
$LIST$
.]
.\"
.TI


NEWS
OpenSS7 -- history of user visible changes.  2015-01-15
$Id: NEWS,v 1.1.2.25 2011-05-10 13:45:30 brian Exp $
Copyright (c) 2008-2015  Monavacon Limited. <http://www.monavacon.com/>
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2001  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
'OpenSS7' package.

Major changes for release openss7-1.1.7.20150115
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Source code release on GitHub.

Major changes for release openss7-1.1.7.20131209
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build corrections for RHEL 6.4.

   - Scripts for country boundary, shoreline and rivers database
     preparation.

   - Database for ANSI T1.101 point code assignments.

   - Suppress extransous /proc/1/comm error messages when testing for
     systemd.

   - Handle PUIAS to Springdale rename for PUIAS distribution.

Major changes for release openss7-1.1.7.20131123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Correction to ca-cert handling in RHEL 6 install scriptlet.

   - RHEL 6 build changes.

   - Set permissions correctly when applying kernel module patch.

   - Do not duplicate certs when including ca-cert.

   - Avoid conflicting shelll variable in RPM install scriplet.

   - Remove /proc/1/comm checks for systemd in module loader.

Major changes for release openss7-1.1.7.20130209
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Corrections to RPM spec file.

Major changes for release openss7-1.1.7.20130129
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Compensate for RHEL6 lack of c_rehash.

   - Do not create sysfs nodes: they do not work properly.

   - Use no foreground flag in strace.service file.

   - More error traces for strerr logger.

   - Remove static device names (the do not work either).

   - Changes to support weak kernel modules.

   - Clean up weak module support.

Major changes for release openss7-1.1.7.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Better Archlinux install script support.

Major changes for release openss7-1.1.6.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Remove use of kill_litter_super().

Major changes for release openss7-1.1.3.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Updates and corrections to weak kernel module builds.

   - Repo support and corrections for Archlinux packages.

Major changes for release openss7-1.1.3.20130123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Additional support for Archlinux.

   - Add udev rules for strace and strerr.

   - Add system unit files for strace and strerr.

   - Add repo support for Archlinux packages.

   - Build Archlinux packages better.

Major changes for release openss7-1.1.3.20130121
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build support for SLES 11.

   - Handle certificates better.

   - Get configuration files in the right places for SLES 11.

   - Support in RPM spec file for certificates on SLES 11.

   - Create a new OpenSS7 certificate.

   - Repository support for SLES 11.

   - RPM spec file reworked for new RPM release.

   - Better certificate handling for RHEL.

   - Additional build support for RHEL 6 with certificates.


src/modules/cd_daed.c
/*****************************************************************************

 @(#) File: src/modules/cd_daed.c

 -----------------------------------------------------------------------------

 Copyright (c) 2008-2015  Monavacon Limited <http://www.monavacon.com/>
 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU Affero General Public License as published by the Free
 Software Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more
 details.

 You should have received a copy of the GNU Affero General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>, or
 write to the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA
 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 *****************************************************************************/

static char const ident[] = "src/modules/cd_daed.c (" PACKAGE_ENVR ") " PACKAGE_DATE;

#include <sys/os7/compat.h>

#include <sys/cdi.h>
#include <sys/cdi_daed.h>
#include <ss7/chi.h>
#include <ss7/chi_ioctl.h>
#include <ss7/lmi.h>
#include <ss7/lmi_ioctl.h>
#include <ss7/hdlc_ioctl.h>

#include "cd/cd.h"

#define CD_DAED_DESCRIP		"Q.703/T1.111.3 DAED: (Delimination Alignment and Error Detection) STREAMS Module"
#define CD_DAED_EXTRA		"Part of the OpenSS7 SS7 Stack for Linux Fast-STREAMS"
#define CD_DAED_COPYRIGHT	"Copyright (c) 2008-2015  Monavacon Limited.  All Rights Reserved."
#define CD_DAED_REVISION	"OpenSS7 src/modules/cd_daed.c (" PACKAGE_ENVR ") " PACKAGE_DATE
#define CD_DAED_DEVICE		"SVR 4.2 MP STREAMS CDI DAED Module for SS7 Channel Devices (DAED)."
#define CD_DAED_CONTACT		"Brian Bidulock <bidulock@openss7.org>"
#define CD_DAED_LICENSE		"GPL"
#define CD_DAED_BANNER		CD_DAED_DESCRIP		"\n" \
				CD_DAED_EXTRA		"\n" \
				CD_DAED_COPYRIGHT	"\n" \
				CD_DAED_REVISION	"\n" \
				CD_DAED_DEVICE		"\n" \
				CD_DAED_CONTACT		"\n"
#define CD_DAED_SPLASH		CD_DAED_DEVICE		" - " \
				CD_DAED_REVISION	"\n"

#ifdef LINUX
MODULE_AUTHOR(CD_DAED_CONTACT);
MODULE_DESCRIPTION(CD_DAED_DESCRIP);
MODULE_SUPPORTED_DEVICE(CD_DAED_DEVICE);
#ifdef MODULE_LICENSE
MODULE_LICENSE(CD_DAED_LICENSE);
#endif				/* MODULE_LICENSE */
#if defined MODULE_ALIAS
MODULE_ALIAS("streams-cd_daed");
#endif
#ifdef MODULE_VERSION
MODULE_VERSION(PACKAGE_ENVR);
#endif
#endif				/* LINUX */

#define CD_DAED_MOD_ID		CONFIG_STREAMS_CD_DAED_MODID
#define CD_DAED_MOD_NAME	CONFIG_STREAMS_CD_DAED_NAME

/*
 *  =======================================================================
 *
 *  STREAMS Definitions
 *
 *  =======================================================================
 */

#define MOD_ID		CD_DAED_MOD_ID
#define MOD_NAME	CD_DAED_MOD_NAME

STATIC struct module_info daed_winfo = {
	mi_idnum:MOD_ID,		/* Module ID number */
	mi_idname:MOD_NAME "-wr",	/* Module name */
	mi_minpsz:(1),			/* Min packet size accepted */
	mi_maxpsz:INFPSZ,		/* Max packet size accepted */
	mi_hiwat:(1),			/* Hi water mark */
	mi_lowat:(0),			/* Lo water mark */
};

STATIC struct module_info daed_rinfo = {
	mi_idnum:MOD_ID,		/* Module ID number */
	mi_idname:MOD_NAME "-rd",	/* Module name */
	mi_minpsz:(1),			/* Min packet size accepted */
	mi_maxpsz:INFPSZ,		/* Max packet size accepted */
	mi_hiwat:(1),			/* Hi water mark */
	mi_lowat:(0),			/* Lo water mark */
};

STATIC streamscall int daed_open(queue_t *, dev_t *, int, int, cred_t *);
STATIC streamscall int daed_close(queue_t *, int, cred_t *);

STATIC struct qinit daed_rinit = {
	qi_putp:ss7_oput,		/* Read put (message from below) */
	qi_srvp:ss7_osrv,		/* Read queue service */
	qi_qopen:daed_open,		/* Each open */
	qi_qclose:daed_close,		/* Last close */
	qi_minfo:&daed_rinfo,		/* Information */
};

STATIC struct qinit daed_winit = {
	qi_putp:ss7_iput,		/* Write put (message from above) */
	qi_srvp:ss7_isrv,		/* Write queue service */
	qi_minfo:&daed_winfo,		/* Information */
};

STATIC struct streamtab cd_daedinfo = {
	st_rdinit:&daed_rinit,		/* Upper read queue */
	st_wrinit:&daed_winit,		/* Upper write queue */
};

STATIC int
daed_init_caches(void)
{
	return (0);
}

STATIC int
daed_term_caches(void)
{
	return (0);
}

/*
 *  =========================================================================
 *
 *  OPEN and CLOSE
 *
 *  =========================================================================
 */
/*
 *  OPEN
 *  -------------------------------------------------------------------------
 */
STATIC struct str *daed_list = NULL;
STATIC streamscall int
daed_open(queue_t *q, dev_t *devp, int flag, int sflag, cred_t *crp)
{
	MOD_INC_USE_COUNT;	/* keep module from unloading in our face */
	if (q->q_ptr != NULL) {
		MOD_DEC_USE_COUNT;
		return (0);	/* already open */
	}
	if (sflag == MODOPEN || WR(q)->q_next != NULL) {
		int cmajor = getmajor(*devp);
		int cminor = getminor(*devp);
		struct str *str;

		/* test for multiple push */
		for (str = daed_list; str; str = str->next) {
			if (str->u.dev.cmajor == cmajor && str->u.dev.cminor == cminor) {
				MOD_DEC_USE_COUNT;
				return (ENXIO);
			}
		}
		if (!(str = cd_alloc_priv(q, &daed_list, devp, crp, CD_DAED))) {
			MOD_DEC_USE_COUNT;
			return (ENOMEM);
		}
		qprocson(q);
		return (0);
	}
	MOD_DEC_USE_COUNT;
	return (EIO);
}

/*
 *  CLOSE
 *  -------------------------------------------------------------------------
 */
STATIC streamscall int
daed_close(queue_t *q, int flag, cred_t *crp)
{
	(void) flag;
	(void) crp;
	cd_free_priv(q);
	MOD_DEC_USE_COUNT;
	return (0);
}

#ifdef LINUX
/*
 *  Linux Registration
 *  -------------------------------------------------------------------------
 */

unsigned short modid = MOD_ID;

#ifndef module_param
MODULE_PARM(modid, "h");
#else
module_param(modid, ushort, 0444);
#endif
MODULE_PARM_DESC(modid, "Module ID for the CD-DAED module. (0 for allocation.)");

/*
 *  Linux Fast-STREAMS Registration
 *  - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 */
STATIC struct fmodsw daed_fmod = {
	.f_name = MOD_NAME,
	.f_str = &cd_daedinfo,
	.f_flag = D_MP,
	.f_kmod = THIS_MODULE,
};

STATIC int
daed_register_strmod(void)
{
	int err;

	if ((err = register_strmod(&daed_fmod)) < 0)
		return (err);
	return (0);
}

STATIC int
daed_unregister_strmod(void)
{
	int err;

	if ((err = unregister_strmod(&daed_fmod)) < 0)
		return (err);
	return (0);
}

MODULE_STATIC int __init
cd_daedinit(void)
{
	int err;

#ifdef MODULE
	cmn_err(CE_NOTE, CD_DAED_BANNER);	/* banner message */
#else
	cmn_err(CE_NOTE, CD_DAED_SPLASH);	/* console splash */
#endif
	if ((err = daed_init_caches())) {
		cmn_err(CE_WARN, "%s: could not init caches, err = %d", MOD_NAME, err);
		return (err);
	}
	if ((err = daed_register_strmod())) {
		cmn_err(CE_WARN, "%s: could not register module, err = %d", MOD_NAME, err);
		daed_term_caches();
		return (err);
	}
	if (modid == 0)
		modid = err;
	return (0);
}

MODULE_STATIC void __exit
cd_daedterminate(void)
{
	int err;

	if ((err = daed_unregister_strmod()))
		cmn_err(CE_WARN, "%s: could not unregister module", MOD_NAME);
	if ((err = daed_term_caches()))
		cmn_err(CE_WARN, "%s: could not terminate caches", MOD_NAME);
	return;
}

/*
 *  Linux Kernel Module Initialization
 *  -------------------------------------------------------------------------
 */
module_init(cd_daedinit);
module_exit(cd_daedterminate);

#endif				/* LINUX */
